### DISCUSSION TITLE
Use of automatically generated mode transition diagrams

### DISCUSSION THREAD

**Discussion by pierreganty** [neovim/neovim#28304] [Ideas]
I've experienced with the automated generation of diagrams showing how keys
transition between modes (insert, normal,…). To sum up, it is possible to
generate this kind of [diagram](https://gist.githubusercontent.com/darcyparker/1886716/raw/c1ee7657010278a787c6502b796a6766a40d56aa/vimModeStateDiagram.svg) with no manual intervention.  

An obvious use case for such diagrams is user documentation, but I think it has
even more potential for developers. As a matter of fact, such diagrams can be
used for conformance testing as follows: given a diagram and a Neovim binary we run
tests (thousands of them) to ensure that the Neovim binary complies with the
behavior specified by the diagram. There are already implementations out there
that carry out conformance testing by generating and running the tests arising 
from a diagram. Such testing strategy could be used, for instance, between 
major releases of the editor.

I presented the automated generation of such diagrams at the [SPIN
conference](https://spin-web.github.io/SPIN2024/program). 

Now, I would like feedback from the community to gauge potential interest.

**Comment by dundargoc** [comment]
This feels like a solution looking for a problem to me. I personally don't think that diagram is that useful as user documentation, and I'm pretty sure we can't use images in our documentation as most terminals don't support that (and neither does neovim as I understand it).

As for the testing: it's very interesting personally as I haven't heard about this methodology. I am curious though if the current testing suite doesn't pretty much already ensure conformance though? What would be the benefit of this that isn't covered by the current testing implementation? Does the benefits outweigh the costs of implementing and maintaining the diagram?

**Comment by pierreganty** [comment]
Fair enough, let's forget about user documentation and focus on testing.

As evidence of "what this brings to the table" let me mention several previous issues that have been reported and fixed thanks to the automatically generated diagrams. The issues were not covered by tests. 

https://github.com/vim/vim/issues/12684

https://github.com/vim/vim/issues/13649

https://github.com/vim/vim/issues/13091

https://github.com/vim/vim/issues/12115

https://github.com/neovim/neovim/issues/15288

https://github.com/vim/vim/issues/8342

**Reply by dundargoc** [reply]
Interesting. Thanks for the examples. If you're interested in pursuing this then I think your best bet is to demonstrate its usefulness in the form of a pull request. I suggest to start minimal (but ideally still functional) PR to really kickstart the discussion.

**Reply by pierreganty** [reply]
I would need to depend on the following python package: [AALpy](https://pypi.org/project/aalpy/). Then the PR should be rather short to implement the conformance testing functionality. Is that a viable path forward ? Or do you already see some red flags ?

**Reply by dundargoc** [reply]
It's not great since we're actively trying to reduce the amount of python dependencies, but I wouldn't say it's an outright dealbreaker. If you don't think it'll take much time then to write a minimal working example I say it's worth a shot to write a PR to test the waters. If you do anticipate that it'll take a lot of time then my gut feeling is to skip it. That's my view on it at least.
